From c17262295b6e3052a1b646d5303326c568b698a0 Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Tue, 13 Oct 2020 23:11:33 +0800
Subject: [PATCH] Cat73 -> Don't tick Skulls - unused code
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

头颅的 tickloop
中只会对两个变量进行更新，然鹅这俩变量是私有的且没在任何地方使用，所以可以安全的移除掉

diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 5528df3d..f8dc19d0 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -24,7 +24,7 @@ import com.mojang.authlib.ProfileLookupCallback;
 import java.util.concurrent.Callable;
 // Spigot end
 
-public class TileEntitySkull extends TileEntity implements ITickable {
+public class TileEntitySkull extends TileEntity /* implements ITickable */ {
 
     @Nullable
     private static UserCache userCache;
@@ -127,7 +127,7 @@ public class TileEntitySkull extends TileEntity implements ITickable {
 
     }
 
-    @Override
+    // @Override
     public void tick() {
         IBlockData iblockdata = this.getBlock();
 
-- 
2.28.0.windows.1

