From da0bae5e1fa7d2d603e881bbebc8a511babff516 Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Tue, 25 Feb 2020 23:21:27 +0800
Subject: [PATCH] Cat73 -> COD Control
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

控制鳕鱼的量，这玩意吃 CPU 比较多

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 72b8677f..192435c3 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -579,6 +579,13 @@ public class CraftEventFactory {
         org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity();
 
         EntitySpawnEvent event = new EntitySpawnEvent(bukkitEntity);
+        // Cat73 start - 让鳕鱼生成有概率失败
+        if (bukkitEntity.getType() == EntityType.COD) {
+            if (Entity.SHARED_RANDOM.nextInt(100) != 0) {
+                event.setCancelled(true);
+            }
+        }
+        // Cat73 end
         bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
-- 
2.25.0

