From ccd57992169ceb01bbeb231fc1d51eb18f9b9ed3 Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Wed, 7 Aug 2019 11:40:44 +0800
Subject: [PATCH] Cat73 -> COD Control
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

控制鳕鱼的量，这玩意在这个版本吃 CPU 比较多

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 67076bdd1..432aee014 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -566,6 +566,13 @@ public class CraftEventFactory {
         org.bukkit.entity.Entity bukkitEntity = entity.getBukkitEntity();
 
         EntitySpawnEvent event = new EntitySpawnEvent(bukkitEntity);
+        // Cat73 start - 让鳕鱼生成有概率失败
+        if (bukkitEntity.getType() == EntityType.COD) {
+            if (Entity.SHARED_RANDOM.nextInt(100) != 0) {
+                event.setCancelled(true);
+            }
+        }
+        // Cat73 end
         bukkitEntity.getServer().getPluginManager().callEvent(event);
         return event;
     }
-- 
2.22.0

