From d954e7ba5cddacbbdd88e3aae82f6fe45d230c4c Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Mon, 12 Aug 2019 18:13:14 +0800
Subject: [PATCH] Cat73 -> debug code


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 825e72e53..400d9554d 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -67,6 +67,7 @@ public class WorldServer extends World {
     private boolean ticking;
     @Nullable
     private final MobSpawnerTrader mobSpawnerTrader;
+    private volatile boolean thisHUse;
 
     // CraftBukkit start
     private int tickPosition;
@@ -1066,12 +1067,20 @@ public class WorldServer extends World {
 
         this.getScoreboard().a(entity);
         if (entity instanceof EntityInsentient) {
+            this.checkH();
             this.H.remove(((EntityInsentient) entity).getNavigation());
         }
 
         entity.valid = false; // CraftBukkit
     }
 
+    private void checkH() {
+        if (this.thisHUse) {
+            MinecraftServer.LOGGER.error("checkH error call stack", new Throwable());
+            MinecraftServer.LOGGER.error("Current Thread Name: " + Thread.currentThread().getName());
+        }
+    }
+
     private void registerEntity(Entity entity) {
         org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot
         if (this.tickingEntities) {
@@ -1092,6 +1101,7 @@ public class WorldServer extends World {
             this.entitiesByUUID.put(entity.getUniqueID(), entity);
             this.getChunkProvider().addEntity(entity);
             if (entity instanceof EntityInsentient) {
+                this.checkH();
                 this.H.add(((EntityInsentient) entity).getNavigation());
             }
             entity.valid = true; // CraftBukkit
@@ -1200,6 +1210,7 @@ public class WorldServer extends World {
         VoxelShape voxelshape1 = iblockdata1.getCollisionShape(this, blockposition);
 
         if (VoxelShapes.c(voxelshape, voxelshape1, OperatorBoolean.NOT_SAME)) {
+            this.thisHUse = true;
             Iterator iterator = this.H.iterator();
 
             while (iterator.hasNext()) {
@@ -1209,6 +1220,7 @@ public class WorldServer extends World {
                     navigationabstract.b(blockposition);
                 }
             }
+            this.thisHUse = false;
 
         }
     }
-- 
2.22.0

