From a1b2d8827a2466bd196a7dce7e8eec637491b6c2 Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Sat, 10 Aug 2019 13:45:41 +0800
Subject: [PATCH] Cat73 -> Disable rcon log


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 58e89dd27..77fddd33c 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1625,7 +1625,8 @@ public class PlayerConnection implements PacketListenerPlayIn {
         org.bukkit.craftbukkit.SpigotTimings.playerCommandTimer.startTiming(); // Spigot
         // CraftBukkit start - whole method
         if ( org.spigotmc.SpigotConfig.logCommands ) // Spigot
-        this.LOGGER.info(this.player.getName() + " issued server command: " + s);
+            if (!java.util.Objects.equals(this.player.getName(), "Rcon")) // Cat73
+                this.LOGGER.info(this.player.getName() + " issued server command: " + s);
 
         CraftPlayer player = this.getPlayer();
 
diff --git a/src/main/java/net/minecraft/server/RemoteControlSession.java b/src/main/java/net/minecraft/server/RemoteControlSession.java
index c28a034fe..2f311a970 100644
--- a/src/main/java/net/minecraft/server/RemoteControlSession.java
+++ b/src/main/java/net/minecraft/server/RemoteControlSession.java
@@ -28,7 +28,7 @@ public class RemoteControlSession extends RemoteConnectionThread {
         }
 
         this.l = s;
-        this.b("Rcon connection from: " + socket.getInetAddress());
+        // this.b("Rcon connection from: " + socket.getInetAddress()); // Cat73
     }
 
     public void run() {
-- 
2.22.0

