From 72304511879a933dbd76b7bd46cd24ef356f660b Mon Sep 17 00:00:00 2001
From: Cat73 <root@cat73.org>
Date: Sun, 10 Mar 2019 22:11:22 +0800
Subject: [PATCH] Cat73 -> Set ServerThread priority to 10
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

提升主线程优先级，由 5(默认, 中等) 提升到 10(最高)，可能能让主线程抢到更多的 CPU

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 72de4043..e814ef87 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -695,6 +695,7 @@ public abstract class MinecraftServer implements IAsyncTaskHandler, IMojangStati
 
     public void run() {
         try {
+            Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
             if (this.init()) {
                 this.nextTick = SystemUtils.getMonotonicMillis();
                 this.m.setMOTD(new ChatComponentText(this.motd));
-- 
2.21.0

